src/isolate.rs

```diff
@@ -283,23 +289,77 @@ impl Isolate {
     }
   }
 
+  /// Runs the default MicrotaskQueue until it gets empty.
+  /// Any exceptions thrown by microtask callbacks are swallowed.
+  pub fn run_microtasks(&mut self) {
+    unsafe { v8__Isolate__RunMicrotasks(self) }
+  }
+
+  /// Enqueues the callback to the default MicrotaskQueue
+  pub fn enqueue_microtask(&mut self, microtask: Local<Function>) {
+    unsafe { v8__Isolate__EnqueueMicrotask(self, microtask) }
+  }
+
+  /// Disposes the isolate.  The isolate must not be entered by any
+  /// thread to be disposable.
+  unsafe fn dispose(&mut self) {
+    IsolateHandle::dispose(self);
+    v8__Isolate__Dispose(self)
+  }
+}
+
+// TODO(ry) Use AtomicPtr? Use Cell? Only one thread ever modifies it.
+/// Thread safe reference to an Isolate. This handle will out-live the Isolate
+/// itself, when that happens all methods on it will return false.
+pub struct IsolateHandle(Arc<Mutex<Option<*mut Isolate>>>);
+
+unsafe impl Send for IsolateHandle {}
+unsafe impl Sync for IsolateHandle {}
+
+impl IsolateHandle {
+  fn dispose(isolate: &mut Isolate) {
+    let slot_ptr = isolate.get_data(0) as *const Mutex<Option<*mut Isolate>>;
+    if !slot_ptr.is_null() {
+      unsafe { isolate.set_data(0, std::ptr::null_mut()) };
+      let handle = unsafe { Arc::from_raw(slot_ptr) };
+      let some_ptr = handle.lock().unwrap().take();
+      assert!(some_ptr.is_some());
+    }
+  }
+
+  fn new(isolate: &mut Isolate) -> Self {
+    let slot_ptr = isolate.get_data(0) as *mut Mutex<Option<*mut Isolate>>;
+    if slot_ptr.is_null() {
+      let x: Option<*mut Isolate> = Some(&mut *isolate);
+      let handle = Arc::new(Mutex::new(x));
+      let handle_ptr = Arc::into_raw(handle.clone());
+      unsafe {
+        isolate.set_data(0, handle_ptr as *mut c_void);
+      }
+      IsolateHandle(handle)
+    } else {
+      let handle = unsafe { Arc::from_raw(slot_ptr) };
+      // Call into_raw so again to avoid double free.
+      let _ptr = Arc::into_raw(handle.clone());
```

Comment:
This is a memory leak, as dropping the IsolateHandle this clone is created for does not decrease the reference count.

Reply:
Nevermind, I think I was mistaken.